red4ext-sys/src/interop.rs

```diff
@@ -46,6 +48,97 @@ unsafe impl ExternType for CName {
     type Kind = cxx::kind::Trivial;
 }
 
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(C)]
+pub struct ResRef(RaRef);
+
+impl ResRef {
+    pub fn new(path: &str) -> Result<Self, ResourcePathError> {
+        Ok(Self(RaRef::new(path)?))
+    }
+}
+
+unsafe impl ExternType for ResRef {
+    type Id = type_id!("RED4ext::ResRef");
+    type Kind = cxx::kind::Trivial;
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(C)]
+pub struct RaRef(ResourcePath);
+
+impl RaRef {
+    fn new(path: &str) -> Result<Self, ResourcePathError> {
+        Ok(Self(ResourcePath::try_new(path)?))
+    }
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct ResourcePath {
+    hash: u64,
+}
+
+impl ResourcePath {
+    const MAX_LENGTH: usize = 216;
+
+    /// accepts non-sanitized path of any length,
+    /// but final sanitized path length must be equals or inferior to 216 bytes
+    fn try_new(path: &str) -> Result<Self, ResourcePathError> {
+        let sanitized = path
+            .trim_start_matches(|c| c == '\'' || c == '\"')
+            .trim_end_matches(|c| c == '\'' || c == '\"')
+            .trim_start_matches(|c| c == '/' || c == '\\')
+            .trim_end_matches(|c| c == '/' || c == '\\')
+            .split(|c| c == '/' || c == '\\')
+            .filter(|comp| !comp.is_empty())
+            .map(str::to_ascii_lowercase)
+            .reduce(|mut acc, e| {
+                acc.push('\\');
+                acc.push_str(&e);
+                acc
+            })
+            .ok_or(ResourcePathError::Empty)?;
+        if sanitized.as_bytes().len() > Self::MAX_LENGTH {
+            return Err(ResourcePathError::TooLong {
+                max: ResourcePath::MAX_LENGTH,
+            });
+        }
+        if let Ok(ref segments) = PathBuf::try_from(&sanitized) {
```

Comment:
I think this is a roundabout way of doing things, we first expect a str and then convert it to Path to check that it is canonical (PathBuf::try_from is infallible by the way, you could write it as Path::from).
If we expect the input to be a path we should just expect `impl AsRef<Path>` and validate that and a cleaner way to validate would be to call .components() and check that they're all Component::Normal, it would also cover other cases we care about like a leading slash, which are not covered here.